src/ConnectedMode/QualityProfiles/QualityProfileDownloader.cs

```diff
@@ -0,0 +1,202 @@
+                UpdateProfile(boundProject, language, qualityProfileInfo);
+
+                var bindingConfiguration = configurationPersister.Persist(boundProject);
+
+                // Create the binding configuration for the language
+                var bindingConfig = await bindingConfigProvider.GetConfigurationAsync(qualityProfileInfo, language, bindingConfiguration, cancellationToken);
+                if (bindingConfig == null)
+                {
+                    logger.WriteLine(string.Format(BindingStrings.SubTextPaddingFormat,
+                        string.Format(BindingStrings.FailedToCreateBindingConfigForLanguage, language.Name)));
+                    return false;
+                }
+
+                bindingConfigs.Add(bindingConfig);
+
+                logger.WriteLine(string.Format(BindingStrings.SubTextPaddingFormat,
+                    string.Format(BindingStrings.QualityProfileDownloadSuccessfulMessageFormat, qualityProfileInfo.Name, qualityProfileInfo.Key, language.Name)));
+            }
+
+            solutionBindingOperation.SaveRuleConfiguration(bindingConfigs, cancellationToken);
+
+            return true;
+        }
+
+        /// <summary>
+        /// Ensures that the bound project has a profile entry for every supported language
+        /// </summary>
+        /// <remarks>If we add support for new language in the future, this method will make sure it's
+        /// Quality Profile is fetched next time an update is triggered</remarks>
+        private void EnsureProfilesExistForAllSupportedLanguages(BoundSonarQubeProject boundProject)
+        {
+            if (boundProject.Profiles == null)
+            {
+                boundProject.Profiles = new Dictionary<Language, ApplicableQualityProfile>();
+            }
+
+            foreach (var language in languagesToBind)
+            {
+                if (!boundProject.Profiles.ContainsKey(language))
+                {
+                    boundProject.Profiles[language] = new ApplicableQualityProfile
+                    {
+                        ProfileKey = null,
+                        ProfileTimestamp = DateTime.MinValue,
+                    };
+                }
+            }
+        }
+
+        private static void UpdateProfile(BoundSonarQubeProject boundSonarQubeProject, Language language, SonarQubeQualityProfile serverProfile)
+        {
+            boundSonarQubeProject.Profiles[language] = new ApplicableQualityProfile
+            {
+                ProfileKey = serverProfile.Key, ProfileTimestamp = serverProfile.TimeStamp
+            };
+        }
+
+        /// <summary>
+        /// Attempts to fetch the QP for the specified language.
+        /// </summary>
+        /// <returns>The QP, or null if the language plugin is not available on the server</returns>
+        private async Task<SonarQubeQualityProfile> TryDownloadQualityProfileAsync(BoundSonarQubeProject boundProject, Language language, CancellationToken cancellationToken)
```

Comment:
This method is unchanged, except that the `BoundSonarQubeProject` is passed in, rather than the equivalent data being extracted from the `BindingArgs` instance.